http-casing/src/com/untangle/app/http/HttpUtility.java

```diff
@@ -0,0 +1,283 @@
+/**
+ * $Id: $
+ */
+package com.untangle.app.http;
+
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLEngine;
+
+import org.apache.hc.core5.net.URIBuilder;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.LogManager;
+
+/**
+ * Utility class for Extract host name
+ */
+public class HttpUtility {
+
+    // these are used while extracting the SNI from the SSL ClientHello packet
+    private static int TLS_HANDSHAKE = 0x16;
+    private static int CLIENT_HELLO = 0x01;
+    private static int SERVER_NAME = 0x0000;
+    private static int HOST_NAME = 0x00;
+
+    public static final String CAPTIVE_PORTAL = "captive-portal";
+    public static final String SSL_INSPECTOR = "ssl-inspector";
+    public static final String THREAD_PREVENTION = "threat-prevention";
+    public static final String WEB_FILTER = "web-filter";
+
+    private static final HttpUtility INSTANCE = new HttpUtility();
+    
+    private static final Logger logger = LogManager.getLogger(HttpUtility.class);
+
+    /**
+     * private constructor
+     */
+    private HttpUtility(){}
+
+    /**
+     * Get the global HttpUtility singleton
+     * @return the INSTANCE
+     */
+    public static HttpUtility getInstance()
+    {
+        return INSTANCE;
+    }
+
+    /**
+     * Function for extracting the SNI hostname from the client request.
+     * @param data
+     * @param appName
+     * @return The SNI hostname extracted from the client request, or null
+     * @throws Exception
+     */
+    public static String extractSniHostname(ByteBuffer data, String appName) throws Exception{
+        int counter = 0;
+        int pos=0;
+        if(CAPTIVE_PORTAL.equals(appName) || SSL_INSPECTOR.equals(appName)){
```

Comment:
Why do we need two separate functions based on app?  I would think that whatever extra work/checking being done for  SSLI and Captive portal are doing would apply to WebFilter and ThreatPrevention as well.

That would considerably simplify the code and remove the need to pass app names.

Reply:
Yes, code has been refactored, please review.

Reply:
A single function has been created instead of two separate apps because an additional check for appName is required. This is necessary due to the differing behaviours in handling exceptions versus null checks for each app. Therefore, we needed to include appName in this implementation.
In SSLinspectorParserEventHandler, if snihostname== null then this is handled along with session.getServerPort() != 25 and checking the

Reply:
In the file SSLinspectorParserEventHandler, the only instance where sniHostname == null is checked is at line 684:if (session.getServerPort() != 25 && sniHostname == null) {
    serverCert = UvmContextFactory.context().certCacheManager().fetchServerCertificate(session.getServerAddr().getHostAddress().toString());
}
Additionally, while throwing an exception, we evaluate rules at line 578 in the same file.

Reply:
I'm not seeing where we need the app name processing.   I'm also concerned about performance in checking the app name on each call.  If we need to any kind of conditional at all, it should be a boolean set by the callers, not their app stirng.

Looking at the code, I see two reasons to check the apps:

1. To avoid throwing an exception for captive portal/ssli in extractSniHostNameFromExtension.  But both of those do handle exceptions and I think they need to continue to get them.
2. Handling "extra" processing that either SSLI or captive portal handled in the hello packet.  I don't think how those are processed would affect either webfilter or threat prevention.

If you 1) remove the app checks in extractSniHostNameFromExtension and always the the exceptions and  2) remove the app checks in extractSniHostname, do all four test suites still pass?  If they do, I recommend going down that path of not checking app names.

Reply:
Done

Reply:
Done. 
Attached is the result of the TC